pkg/tcpip/stack/stack.go

```diff
@@ -1161,17 +1168,32 @@ func (s *Stack) FindRoute(id tcpip.NICID, localAddr, remoteAddr tcpip.Address, n
 				continue
 			}
 
+			plen := route.Destination.Mask().Prefix()
+			if plen < bestPrefixLen {
+				continue
+			}
+
 			if id == 0 || id == route.NIC {
 				if addressEndpoint := s.getAddressEP(nic, localAddr, remoteAddr, netProto); addressEndpoint != nil {
 					var gateway tcpip.Address
 					if needRoute {
 						gateway = route.Gateway
 					}
-					r := constructAndValidateRoute(netProto, addressEndpoint, nic /* outgoingNIC */, nic /* outgoingNIC */, gateway, localAddr, remoteAddr, s.handleLocal, multicastLoop)
-					if r == nil {
+
+					if r := constructAndValidateRoute(netProto, addressEndpoint, nic /* outgoingNIC */, nic /* outgoingNIC */, gateway, localAddr, remoteAddr, s.handleLocal, multicastLoop); r == nil {
+						panic(fmt.Sprintf("non-forwarding route validation failed with route table entry = %#v, id = %d, localAddr = %s, remoteAddr = %s", route, id, localAddr, remoteAddr))
+					} else if plen >= bestPrefixLen {
+						bestRoute, bestPrefixLen = r, plen
+					}
+				} else {
+					// There's no local address on the NIC but this route suggests that
+					// we can still reach this destination address via this NIC, so just
+					// construct a route that uses the remote address as the gateway.
+					if r := constructAndValidateRoute(netProto, nil, nic /* outgoingNIC */, nic /* outgoingNIC */, remoteAddr, localAddr, remoteAddr, s.handleLocal, multicastLoop); r == nil {
```

Comment:
shouldn't we always pass an emty value for localAddr here?

Reply:
As of the latest commit, we will pass `bestLocalAddr` which will be empty if not populated with a closer source address.

Reply:
what is "closer source address"

If i'm parsing this CL correctly, `bestLocalAddr` will only ever be `""` or `localAddr`

am i missing something?